[UI Tests] Add Domains Dashboard Card UI test

pachlava opened the pull request:
### Description
This PR adds a UI test for the Domains Dashboard card:

<img width="275" alt="Screenshot 2023-04-19 at 15 15 16" src="https://user-images.githubusercontent.com/73365754/233071842-ae66eaf6-e3ad-4b0c-bdc8-3f0eeb5a0803.png">

What it does:
- Checks that that card with expected header and text is available at `Home` screen
- Taps the card and checks that the redirect to `Domains` screen takes place

I hope the code is pretty self-explanatory, so I'm not providing a lot of notes on it.

### To test
- All tests are green on CI, including the new `testDomainsCardNavigation`

### Regression Notes
I'm not filling the regression notes since the PR adds a UI test, it does not change the app behaviour in any way.


### P.S.
To have less noise here, I'll follow up this PR with another one that fixes / adds missing mocks files for `Home` screen, to avoid seeing this:

<img width="287" alt="Screenshot 2023-04-19 at 15 18 34" src="https://user-images.githubusercontent.com/73365754/233072828-aa8089ab-42eb-4eb1-a142-9baba82380ce.png">

and this:

<img width="1085" alt="Screenshot 2023-04-19 at 15 44 16" src="https://user-images.githubusercontent.com/73365754/233078481-011f2aa4-f5bc-4ddd-b42a-e3eb986c4c04.png">

This does not affect the current test, but is not so pleasant to the eye.

pachlava committed:
First iteration of `testDomainsCard` test

pachlava committed:
Added a mock file for `feature-flags` and renamed a test

pachlava committed:
Trying a rerun to rebuild JP prototype build

pachlava commented on API-Mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/feature-flags.json:
```diff
@@ -0,0 +1,17 @@
+  "response": {
+    "status": 200,
+    "jsonBody": {
+      "dashboard_card_domain": true
```
Mocking this response causes the card to be shown for eligible sites. Note that this is an app-specific request, not a website-specific. It's sent on the app launch.

jostnes replied:
question: what would happen if the feature flag is removed? should we ensure that in the PR that removes the feature flag (not sure if it's removed via an app code change or not since it looks like it's server side) this mock file/line to turn on feature flag is removed too? wdyt?

pachlava replied:
I assume that the feature flag is removed in the app (but @staskus knows better for sure), and once this happens, this mock file won't be necessary any longer. I'm not sure if having it will harm, though. The same `/wpcom/v2/mobile/feature-flags` request matcher might be useful for testing the future feature flags, so we'd probably remove just the line.

staskus replied:
I agree, I think the file itself is and will be useful in the future. 

If we remove the feature flag from this feature, we will be able to remove it from here. However, I think the tests will continue passing so we can easily forget to make the change on UI tests target. That's not a critical issue but likely to happen.

jostnes commented on WordPress/UITests/Tests/DashboardTests.swift:
```diff
@@ -0,0 +1,28 @@
+            .verifyDomainsCard()
+            .tapDomainsCard()
+
+        XCTAssertTrue(DomainsScreen.isLoaded(), "\"Domains\" screen isn't loaded after \"Domains\" card tap.")
```
what do you think about moving this to a `verifyDomainsScreenLoaded` function so we could chain it in the test? i see it's not the standard way for tests in the repo (i think i got too used to seeing it on WCiOS 😄 ) but i think the test reads better that way

pachlava replied:
Good suggestion @jostnes 👍 Done in 144637bd61211d7a62532eeca7d70f2b5a68a6c3.

jostnes commented on WordPress/UITests/Tests/DashboardTests.swift:
```diff
@@ -0,0 +1,28 @@
+        removeApp()
+    }
+
+    func testDomainsCardNavigation() throws {
```
i ran the test locally on the WP test plan and noticed that it failed, which i think is expected? (is this feature Jetpack only?) perhaps it's time we add some identifiers on the tests to know which are JP-specific and WP-specific? (and to disable the test that shouldn't run on the test plan level)

pachlava replied:
Good point. FWIW, I tried selecting `WordPressUITests` target yesterday for execution, for the same reason - to exclude the test form there - and failed to have the target tests enabled. I assumed that the WP tests were all disabled (and also they do not run on CI), but I was wrong, apparently. 

I could do this now, though, it looks like the process is now different after WP/JP separation.

I excluded the test from `WordPressUITests.xctestplan` and added a comment above the test with 71dbf793c6d5b9cd73cf5564fb6c71df0046a4c2

pachlava replied:
> is this feature Jetpack only?

Yes, it's JP only 👍

staskus replied:
I'm not sure if this separation of UI tests was discussed before. I haven't worked much on UI tests.

Theoretically we could make a distinction between Jetpack and WordPress tests even more clearly:

- Add them to separate `JetpackUITests` folder
- Add them to only to `JetpackUITests` target
- Name differently (?), e.g `JetpackDashboardTests`, although it may be unecessary

<img width="229" alt="image" src="https://user-images.githubusercontent.com/4062343/233369746-00b98bc8-111e-4cf8-a30f-cbebc73edc56.png">

What are your thoughts on that? What would be the most comfortable way?

pachlava replied:
In current repo, it's done through Test Plans. For `Jetpack` scheme, the associated Test Plan is `JetpackUITests`, and for `WordPress` it's `WordPressUnitTests` as default one, with a possibility to select `WordPressUITests` Test Plan.

This is different from what how it was the last time I worked with this repo, before the tests were separated between WP and JP. Then, we had to select a corresponding `...UITests` scheme to run the tests.

We have a possibility to activate/deactivate test for a separate Test Plan, and from my POV, this is quite a comfortable way to manage this. I just did not do this for this test, because I mistakenly thought that all UI tests were now disabled for WP.

@jostnes and @tiagomar worked on separating the tests, so maybe they'll share more context here.

staskus replied:
> ...and from my POV, this is quite a comfortable way to manage this

Got it! Test plans definitely make sense! 👍

jostnes reviewed (commented):
thanks for adding this so quickly @pachlava 💯 

i added a couple of comments/questions but one that i think should be addressed is to handle the test failing when running locally against WP (think we should disable the test for WP). 

for the test identifier, i'm not sure which is better, by adding a comment on top of the test or by adding a prefix to the test name or maybe something else so that we clearly know it's for JP only?

pachlava committed:
Disabled the test for WP.

pachlava committed:
Moved Domains screen assertion to the screen file.

staskus commented on WordPress/UITests/Tests/DashboardTests.swift:
```diff
@@ -0,0 +1,28 @@
+    // This test is JP only.
+    func testDomainsCardNavigation() throws {
+        try MySiteScreen()
+            .verifyDomainsCard()
```
A dashboard can have more cards and elements than fit into view and it may require scrolling.

Do we want to ensure we handle such cases or only come back to it when we write tests for more dashboard cards? WDYT?

pachlava replied:
That's a very good point - I'll work on adding a scroll and will let you know once it's done (in current PR). Thanks 🙇

pachlava replied:
@staskus 👋 

I added a scroll to card with 6daeaab698fd3cff218929c7d3e8f0c90283c108. To have it working, I needed to add an accessibility identifier to `DashboardDomainsCardCell.swift` - could you please check if everything is alright with this addition?

Since it's impossible to see if the scroll is working, because the card is fully visible anyway, I made a local test with another card ("Posts") as a sort of proof. The only difference is that I'm referring to the identifier of another cell in the video:

https://user-images.githubusercontent.com/73365754/233623927-b0255c15-0131-40d3-94a1-bbd551c4c309.mov

staskus replied:
Thanks! Yes, it looks good! 👍

staskus commented on WordPress/UITestsFoundation/Screens/MySiteScreen.swift:
```diff
@@ -170,4 +178,18 @@ public class MySiteScreen: ScreenObject {
+
+    @discardableResult
+    public func verifyDomainsCard() -> Self {
+        let cardText = app.staticTexts["Stake your claim on your corner of the web with a site address that’s easy to find, share and follow."]
```
I'm wondering if there's a value in the UI test failing if we decide to change the copy?

IMHO, it's important that such a card appears, it's clickable, and the new DomainsScreen appears. If you agree, maybe we could introduce an identifier that could be used for verifying the card appearing but the success of the test would be independent of the copy. What do you think?

pachlava replied:
Good question! I'd say it depends on how often the copy might change after #20570.

IMO, there's some value in checking that the text is not truncated and that it's displayed on screen (since we check for `isHittable` property, not `exists`).

Using an identifier would make things easier, and I think it makes a lot of sense if we expect the copy to change often.

staskus replied:
We're not planning many changes after https://github.com/wordpress-mobile/WordPress-iOS/pull/20570 but there's a possibility of it changing once in a while. 

> IMO, there's some value in checking that the text is not truncated and that it's displayed on screen (since we check for isHittable property, not exists).

Good point. I'll leave the decision to you then, I don't have a super-strong opinion on this. 

⚠️  I merged #20570 now, so the copy has changed from `Own your online identity with a custom domain` to a shorter `Find a custom domain`

pachlava replied:
I updated from `trunk` and adjusted the copy in d817a7de1a691e6060304e00f7c9281779ff62ba. Waiting for 🟢 CI 🤞

staskus reviewed (commented):
Thank you for the tests! 

I left some questions. I haven't worked much on UI tests in this app, so some questions are more related to general conventions. Thanks!

pachlava committed:
Added an accessibilityIdentifier and a scroll to domains card

pachlava committed:
Merge branch 'trunk' into add-domains-card-ui-test

pachlava committed:
Updated the expected Domains Card copy

pachlava commented:
@jostnes and @staskus, thanks for your reviews 🙇  I made some adjustments, mainly some refactoring, added a scroll to the card before asserting it, and updated the expected card copy (sice it was changed in #20570) - could you please take another look? Thank you!

staskus reviewed (approved):
Thanks for the work and improvements, looks good! 🎉
